"socket_ref reinterpret as void*" test breaks string aliasing rules

```
$ mkdir build
$ cd build/
$ cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_CXX_FLAGS=-Wall -DCMAKE_C_FLAGS=-Wall ..
$ make
```
generates this warning:
```
../cppzmq/tests/socket_ref.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____14()’:
../cppzmq/tests/socket_ref.cpp:112:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  112 |     reinterpret_cast<SVP *>(&ssr)->p = socket.handle();
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
I believe that without `-fstring-aliasing`, this isn't immediately a problem, but depending on future compiler optimizations, maybe it could be.

--- a/tests/socket_ref.cpp
+++ b/tests/socket_ref.cpp
@@ -95,12 +95,12 @@ TEST_CASE("socket_ref swap", "[socket_ref]")
     swap(sr1, sr2);
 }
 
-TEST_CASE("socket_ref reinterpret as void*", "[socket_ref]")
+TEST_CASE("socket_ref type punning", "[socket_ref]")
 {
     struct SVP
     {
         void *p;
-    };
+    } svp;
     struct SSR
     {
         zmq::socket_ref sr;
@@ -109,7 +109,9 @@ TEST_CASE("socket_ref reinterpret as void*", "[socket_ref]")
     zmq::context_t context;
     zmq::socket_t socket(context, zmq::socket_type::router);
     CHECK(socket.handle() != nullptr);
-    reinterpret_cast<SVP *>(&ssr)->p = socket.handle();
+    svp.p = socket.handle();
+    // static_cast to silence incorrect warning
+    std::memcpy(static_cast<void*>(&ssr), &svp, sizeof(ssr));
     CHECK(ssr.sr == socket);
 }
 

